Add support for room sensor accessories assigned to a Honeywell (Lyric) Thermostat

dalinicus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
For the Lyric integration, adds additional sensors for room sensor accessories assigned to a thermostat.  Each room sensor accessory will have a device created for it, sensors created for temperature and humidity, and `via_device` assigned to the respective thermostat.

![image](https://github.com/home-assistant/core/assets/8620219/3f151d43-3942-4866-8af8-fcec5ac1862b)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: n/a
- This PR is related to issue: n/a
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29956

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Commit:
Add support for room sensor accessories.

- Update coordinator to refresh and grab information about room sensor accessories assigned to a thermostat
- Add sensor entities for room humidity and room temperature
- Add devices to the registry for each room accessory
- "via_device" these entities through the assigned thermostat.

Commit:
fixed pre-commit issues.

joostlek commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -158,8 +165,43 @@ class LyricDeviceEntity(LyricEntity):
+        )
+
+
+class LyricAccessoryEntity(LyricDeviceEntity):
```
One question that I have, currently an Accessory is bound to a room. Are there more types of accessories? Are all accessories bound to a room?

dalinicus replied:
I'm not aware of any other thermostat accessories for this device, but they all appear to have the same data structure from the Lyric API.  The only values I have observed for accessories are `Thermostat` (which is a duplicate of the parent thermostat) and `IndoorAirSensor`, and looking at their product pages I can only find the one model of air sensor and no other accessories. `IndoorAirSensor` is the only value they use from within their API reference as well.

Air sensors must assigned a "room" from the Honeywell app in order to be operable, yes.

joostlek commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -116,6 +132,30 @@ class LyricSensorEntityDescription(
+        device_class=SensorDeviceClass.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda _, accessory: accessory.temperature,
+        suitable_fn=lambda _, accessory: accessory.temperature
+        if accessory.type == "IndoorAirSensor"
+        else False,
```
```suggestion
        suitable_fn=lambda _, accessory: accessory.type == "IndoorAirSensor",
```

joostlek replied:
Unless there is a reason why `accessory.temperature` isn't Trueish in some cases. This also prevents a bug where the sensor isn't added if the accessory temperature is exactly 0 (since translating that to a bool makes it false)

dalinicus replied:
Nice catch!  This has been fixed in my most recent commit.

joostlek commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -116,6 +132,30 @@ class LyricSensorEntityDescription(
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda room, _: room.roomAvgHumidity,
+        suitable_fn=lambda room, accessory: room.roomAvgHumidity
+        if accessory.type == "IndoorAirSensor"
+        else False,
```
```suggestion
        suitable_fn=lambda _, accessory: accessory.type == "IndoorAirSensor",
```

dalinicus replied:
Nice catch!  This has been fixed in my most recent commit.

dalinicus commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -158,8 +165,43 @@ class LyricDeviceEntity(LyricEntity):
+        location: LyricLocation,
+        device: LyricDevice,
+        room: LyricRoom,
+        accessory: LyricAccessories,
```
Its singular.  Looks like its misnamed in the aiolyric library that his object comes from.

dalinicus commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -158,8 +165,43 @@ class LyricDeviceEntity(LyricEntity):
     def device_info(self) -> DeviceInfo:
         """Return device information about this Honeywell Lyric instance."""
         return DeviceInfo(
+            identifiers={(dr.CONNECTION_NETWORK_MAC, self._mac_id)},
```
Correct.

Commit:
PR suggestions

- update docstring to reflect ownership by thermostat
- fixed potential issue where a sensor would not be added if its temperature value was 0

dalinicus commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -158,8 +165,43 @@ class LyricDeviceEntity(LyricEntity):
+
+
+class LyricAccessoryEntity(LyricDeviceEntity):
+    """Defines a Honeywell Lyric accessory entity."""
```
Good call. Clarification has been added.

dalinicus commented:
Good Morning (or whatever greeting appropriate for your time zone) @emontnemery and @joostlek ,

I have made and tested the requested code changes, and answered the questions you posed.  I have re-marked the PR for review.  Please don't hesitate to ask followup questions or recommend further changes.   Appreciate all your help on this.

t3hk0d3 commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -76,6 +77,12 @@ async def async_update_data(force_refresh_token: bool = False) -> Lyric:
+                for location in lyric.locations:
+                    for device in location.devices:
+                        if device.deviceClass == "Thermostat":
+                            await lyric.get_thermostat_rooms(
```
Maybe it would be more performant to await all promises altogether?

https://docs.python.org/3/library/asyncio-task.html#asyncio.gather

disforw commented:
@dalinicus nice! Any possibility you can grab the motion as a binary_sensor as well??? Pwwease 🙏

dalinicus commented:
> @dalinicus nice! Any possibility you can grab the motion as a binary_sensor as well??? Pwwease 🙏

Yup! Have it mostly coded but requires the addition of the binary_sensor platform which is new to this integration.  As soon as this one is approved and merged, I'll add that in a seco d pull request cert shortly after

disforw commented:
Sweet deal. I will be looking out for it!
Happy holidays mate..

dalinicus commented:
@emontnemery and @joostlek, figured I'd check in with y'all since its been a month.  Any additional feedback on this change?  I'd be looking to make additional changes to add motion sensor support once this base functionally has been merged. 

Hope you're both having a happy holidays.

Commit:
Merge branch 'dev' into feature/lyric-room-sensor-accessories

joostlek commented:
Wait was this now ready for review or not

disforw commented:
I believe it was ready 👍

Commit:
Merge branch 'dev' into feature/lyric-room-sensor-accessories

Commit:
fix bad github merge

emontnemery commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -77,6 +78,12 @@ async def async_update_data(force_refresh_token: bool = False) -> Lyric:
         try:
             async with asyncio.timeout(60):
                 await lyric.get_locations()
+                for location in lyric.locations:
+                    for device in location.devices:
+                        if device.deviceClass == "Thermostat":
+                            await lyric.get_thermostat_rooms(
+                                location.locationID, device.deviceID
+                            )
```
I agree with @t3hk0d3, please use asyncio.gather instead, something like:
```suggestion
                asyncio.gather(
                    lyric.get_thermostat_rooms(
                        location.locationID, device.deviceID
                    )
                    for device in location.devices
                    for location in lyric.locations
                    if device.deviceClass == "Thermostat"
                )
```

dalinicus replied:
Updated per suggestion; thanks for the code snippet.  Re-tested locally and it continues to work as expected.

emontnemery reviewed (changes requested):
Looks good, just a minor comment.

Commit:
asyicio.gather futures for updating theromstat room stats

emontnemery reviewed (approved):
LGTM, thanks @dalinicus 👍
